### ISSUE TITLE
Options does not change anything

### ISSUE THREAD

**Issue by Kitty-sunray** [yuval-alaluf/restyle-encoder#38]
Hello!
There is set of options available to change the encoder behavior. They got printed out when colab notebook is run:
{'batch_size': 8,
 'board_interval': 50,
 'checkpoint_path': '',
 'dataset_type': 'ffhq_encode',
 'device': 'cuda:0',
 'encoder_type': 'BackboneEncoder',
 'exp_dir': '',
 'id_lambda': 0,
 'image_interval': 100,
 'input_nc': 6,
 'l2_lambda': 1.0,
 'learning_rate': 0.0001,
 'lpips_lambda': 0,
 'max_steps': 500000,
 'moco_lambda': 0,
 'n_iters_per_batch': 5,
 'optim_name': 'ranger',
 'output_size': 1024,
 'save_interval': 10000,
 'start_from_latent_avg': True,
 'stylegan_weights': '',
 'test_batch_size': 8,
 'test_workers': 8,
 'train_decoder': False,
 'val_interval': 10000,
 'w_norm_lambda': 0.0,
 'workers': 8}

Apparently, that ones that supposed to change encoder behavior do not change anything. I tried
 'encoder_type', 'id_lambda',  'l2_lambda', 'lpips_lambda', 'moco_lambda'
Encoder just provides the same output even with lpips, id and moco set to 0 at same time (used ffhq .pt) .

I tried to walk through source code and only found these options been mentioned in the encoder "train" section.
Does that mean we can't control encoder behavior? (I personally want to disable or lower lpips lambda)
What does "training an encoder " means? I just want to embed aligned image into stylegan2 latent space, do not need to "train" anything for it, or do I?

How can we have these options working? I am using colab notebook provided with code.

**Comment by yuval-alaluf** [comment]
When you are using the Colab notebook, you are running inference. The options that are printed are the options that were used to train the encoder. Therefore, changing these values won't do anything because the notebook is performing inference. 
For example, the `lambda` values are the loss weights that are used during training and hence are not relevant for the inference notebook.

To answer your specific questions: 
> Does that mean we can't control encoder behavior? (I personally want to disable or lower lpips lambda)

After the encoder is trained, there are not many options you can control. Basically, the only option for inference would be the number of iterations to perform which can be set using `n_iters_per_batch`. We recommend using `n_iters_per_batch=5`. Anything more than `10` will not help. 

> I just want to embed aligned image into stylegan2 latent space, do not need to "train" anything for it, or do I?

If you want to embed images into the latent space, you can use the notebook for it (or the `inference_iterative` script). You do not need to train anything since the encoder was already trained.

**Comment by Kitty-sunray** [comment]
Thank you for the quick detailed response. It took me longer to sign up here than waiting for your answer!
I got that it is not a bug, sorry for the opened issue.
I was using the https://github.com/robertluxemburg/stylegan2encoder previously, and as far as I (10y iOS dev meeting AI world) can understand, it tries to find best latent code estimating using loss&penalties for each generated image. Since all the loss and penalties can be controlled, it allowed amazing results (i.e. using image not belonging to ffhq, but resembling a human face). But it was clearly lacking of what you implemented. With your work, I instantaneously got very good results. In fact, too good: since image I try to embed is not exactly a photo but more of a art, your encoder tries it best to resemble it in every detail, loosing the realistic "photo" appearance. I was seeking to control it the way I did with robertluxemburg work, but now I understood it is not possible due to encoder already trained with specific losses set. "Training an encoder" is new for me, I used to train generator/discriminator pairs only. And I very like to learn on this, I will go to re-read your paper now. But if you maybe have a quick link to direct me where to read about what training an encoder, or can phrase the purpose of, and a way to replicate such training if few sentences it would be awesome!

**Comment by yuval-alaluf** [comment]
Happy to help :)

Basically, the work you used up until now performed optimization to perform the embedding. With optimization, you do have control over the loss values for performing the gradient descent.  
As you noticed with optimization, each image probably took about a minute or two, but you got good results. Since optimization takes a very long time, it's hard to work with it at large scale. 
With an encoder (what I propose with ReStyle), we actually train a network to take the image and find the corresponding embedding. Since we're using a network to get the embedding (and do not need to perform optimization), the embedding (at inference) takes less than a second. However, the results are typically a bit less accurate then optimization (especially on challenging images). You may also notice that for training the encoder, you do have control over the loss weights and other parameters). 

If you want to learn more about encoders, how to train them, and their relation to optimization techniques, I definitely recommend reading my paper. I hope it helps answer all your questions :) 

As for training an encoder, this repo provides everything you need (more or less) for training your own encoder: https://github.com/yuval-alaluf/restyle-encoder#training

**Comment by Kitty-sunray** [comment]
I got the idea now, bruteforce while juggling parameters VS trained network to check specific latents (and repeat). 
But nothing stops me from feeding latent code found with your AI approach as input to "bruteforce optimizer" with loss functions configured the way I like to fine-tune the final result, right? Sound like a bomb.
Reading paper slowly right now.
off-topic: it is impressive to see such a pro having the goddamn time to write detailed answers. Write a book on time management if you haven't already!

**Comment by yuval-alaluf** [comment]
> But nothing stops me from feeding latent code found with your AI approach as input to "bruteforce optimizer" with loss functions configured the way I like to fine-tune the final result, right? Sound like a bomb.

Exactly. Starting the bruteforce optimizer with the latent code found from the encoder is actually called a "hybrid approach" since it combines the two techniques. Since the initialization is much better, the optimization converges much faster (maybe less than 30 seconds instead of several minutes)

> off-topic: it is impressive to see such a pro having the goddamn time to write detailed answers. Write a book on time management if you haven't already!

Thanks :) 
I always love seeing people interested in my work and asking questions so I am more than happy to take a couple of minutes to help as much as I can! It's fun to see new people getting into the field!
